alica_engine/src/engine/Assignment.cpp

```diff
@@ -1,532 +1,448 @@
-{
-    auto ret = std::make_shared<list<const supplementary::AgentID*>>(list<const supplementary::AgentID*>());
-    // if (this->plan->getEntryPoints().find(ep->getId()) != this->plan->getEntryPoints().end())
-    {
-        auto robots = this->epRobotsMapping->getRobotsByEp(ep);
-        for (int i = 0; i < static_cast<int>(robots->size()); ++i) {
-            ret->push_back(robots->at(i));
-        }
-        for (auto r : (*this->epSucMapping->getRobots(ep))) {
-            if (std::find_if(ret->begin(), ret->end(), [&r](const supplementary::AgentID* id) { return *r == *id; }) == ret->end()) {
-                ret->push_back(r);
-            }
-        }
-    }
-    return ret;
+AssignmentView Assignment::getAgentsWorking(const EntryPoint* ep) const
+{
+    return AssignmentView(this, ep->getIndex());
 }
 
-/**
- * The robots that are currently working on or already succeeded in a specific task, referred to by an EntryPoint Id.
- * @param epid EntryPoint id
- * @return a shared_ptr of a list of int
- */
-std::shared_ptr<list<const supplementary::AgentID*>> Assignment::getRobotsWorkingAndFinished(int64_t epid)
+AssignmentView Assignment::getAgentsWorking(int idx) const
 {
-    auto ret = std::make_shared<std::list<const supplementary::AgentID*>>();
-    auto robots = this->epRobotsMapping->getRobotsByEpId(epid);
-    if (robots != nullptr) {
-        for (int i = 0; i < static_cast<int>(robots->size()); i++) {
-            ret->push_back(robots->at(i));
-        }
-    }
+    return AssignmentView(this, idx);
+}
 
-    auto robotsSucc = this->epSucMapping->getRobotsById(epid);
-    if (robotsSucc != nullptr) {
-        for (auto& robot : *robotsSucc) {
-            ret->push_back(robot);
-        }
-    }
+AssignmentView Assignment::getAgentsWorking(int64_t epId) const
+{
+    const EntryPoint* ep = _plan->getEntryPointByID(epId);
+    return ep ? AssignmentView(this, ep->getIndex()) : AssignmentView();
+}
 
-    return ret;
+AssignmentSuccessView Assignment::getAgentsWorkingAndFinished(const EntryPoint* ep) const
+{
+    return AssignmentSuccessView(this, ep->getIndex());
 }
 
-std::shared_ptr<SuccessCollection> Assignment::getEpSuccessMapping() const
+AllAgentsView Assignment::getAllAgents() const
 {
-    return this->epSucMapping;
+    return AllAgentsView(this);
+}
+AgentsInStateView Assignment::getAgentsInState(const State* s) const
+{
+    return AgentsInStateView(this, s);
 }
 
-void Assignment::setAllToInitialState(const AgentGrp& robots, const EntryPoint* defep)
+AgentsInStateView Assignment::getAgentsInState(int64_t sid) const
 {
-    auto rlist = this->epRobotsMapping->editRobotsByEp(defep);
-    for (const supplementary::AgentID* r : robots) {
-        rlist->push_back(r);
-        robotStateMapping->setState(r, defep->getState());
-    }
+    const State* s = _plan->getStateByID(sid);
+    return s ? AgentsInStateView(this, s) : AgentsInStateView();
```

Comment:
Same comment here as above. Since state ids are static, and because this is potentially called quite often, I might just assume that the id passed into this function MUST exist, so we can get rid of the unnecessary check and not have the user do further sanity checking once they receive the empty AgentsInStateView. I think if an invalid id is passed in here, we should fail.

Then we might not need the default constructor for AgentsInStateView, and this function can simply be: return getAgentsInState(_plan->getStateByID(sid));

Reply:
Cases where calls are quite often should cache the state pointer by themselves. And the default constructor is used in other cases as well.